File: cmd/shengen/brand_inference_test.go
package main

import (
	"os"
	"path/filepath"
	"strings"
	"testing"
)

// brandTableFor parses a spec string and returns the inferred table.
func brandTableFor(t *testing.T, spec string) *BrandTable {
	t.Helper()
	types, err := parseFile_string(spec)
	if err != nil {
		t.Fatal(err)
	}
	st := newSymbolTable()
	st.Build(types)
	return InferBrands(types, st)
}

// brandTableForFile parses a spec on disk (used for the golden tables
// of the committed example specs).
func brandTableForFile(t *testing.T, path string) *BrandTable {
	t.Helper()
	data, err := os.ReadFile(path)
	if err != nil {
		t.Skipf("spec not readable (%v)", err)
	}
	return brandTableFor(t, string(data))
}

func assertGolden(t *testing.T, got, want string) {
	t.Helper()
	normalize := func(s string) string {
		var out []string
		for _, line := range strings.Split(strings.TrimSpace(s), "\n") {
			if l := strings.Join(strings.Fields(line), " "); l != "" {
				out = append(out, l)
			}
		}
		return strings.Join(out, "\n")
	}
	if normalize(got) != normalize(want) {
		t.Errorf("brand table mismatch\n--- got ---\n%s\n--- want ---\n%s", got, want)
	}
}

// ============================================================================
// Golden brand tables for the committed example specs
// ============================================================================

// TestBrandTablePaymentGolden pins the inferred brand table for
// examples/payment/specs/core.shen.
//
// The load-bearing rows:
//
//   - transaction is `minted`: nothing inside it is branded, but
//     balance-invariant and safe-transfer reason about it, so it is the
//     head of a proof chain and mints its own brand.
//   - balance-checked is `inherited` at transaction's brand — the proof
//     is about that transaction, not about transactions in general.
//   - safe-transfer's two premises (Tx, Check) unify, because
//     balance-checked reaches transaction along a nested field path. One
//     surviving brand, so NewSafeTransfer demands a matching pair.
//   - amount / account-id stay unbranded: they are values, not evidence.
//   - account-state stays unbranded: no rule consumes it and none of
//     its constituents is branded.
func TestBrandTablePaymentGolden(t *testing.T) {
	bt := brandTableForFile(t, filepath.Join("..", "..", "examples", "payment", "specs", "core.shen"))
	assertGolden(t, bt.String(), `
account-id               unbranded  AccountId
account-state            unbranded  AccountState  premises(- -)
amount                   unbranded  Amount
balance-checked          inherited  BalanceChecked[B]  premises(- B)
safe-transfer            inherited  SafeTransfer[B]  premises(B B)
transaction              minted     Transaction[B]  premises(- - -)
`)
}

// TestBrandTableMultiTenantGolden pins the inferred brand table for
// examples/multi-tenant-api/specs/core.shen — the full authorization
// chain. One brand threads the whole chain:
//
//	parsed-claims (minted) → verified-jwt → authenticated-user
//	  → human-principal ─┐
//	                     ├→ authenticated-principal (sum)
//	  service-credential (minted) → service-principal ─┘
//	  → tenant-access → resource-access
//
// So a ResourceAccess[B] is evidence about the *specific* JWT that B
// came from, and cannot be paired with another chain's TenantAccess.
func TestBrandTableMultiTenantGolden(t *testing.T) {
	bt := brandTableForFile(t, filepath.Join("..", "..", "examples", "multi-tenant-api", "specs", "core.shen"))
	assertGolden(t, bt.String(), `
authenticated-principal  sum        AuthenticatedPrincipal[B]
authenticated-user       inherited  AuthenticatedUser[B]  premises(B -)
human-principal          inherited  HumanPrincipal[B]  premises(B)
jwt-audience             unbranded  JwtAudience
jwt-issuer               unbranded  JwtIssuer
parsed-claims            minted     ParsedClaims[B]  premises(- - - -)
resource-access          inherited  ResourceAccess[B]  premises(B - -)
resource-id              unbranded  ResourceId
service-credential       minted     ServiceCredential[B]  premises(- -)
service-id               unbranded  ServiceId
service-principal        inherited  ServicePrincipal[B]  premises(B)
tenant-access            inherited  TenantAccess[B]  premises(B - -)
tenant-id                unbranded  TenantId
user-id                  unbranded  UserId
verified-jwt             inherited  VerifiedJwt[B]  premises(B -)
`)
}

// ============================================================================
// Rule-level unit tests
// ============================================================================

// TestBrandUnificationViaFieldPath is the safe-transfer shape reduced to
// its essentials: two premises, one of which structurally contains the
// other, must share a brand.
func TestBrandUnificationViaFieldPath(t *testing.T) {
	bt := brandTableFor(t, `
(datatype thing
  X : string;
  ==============
  [X] : thing;)

(datatype proof
  T : thing;
  Ok : boolean;
  (= Ok true) : verified;
  ==============
  [T Ok] : proof;)

(datatype paired
  T : thing;
  P : proof;
  ==============
  [T P] : paired;)
`)
	if got := bt.Params("thing"); len(got) != 1 {
		t.Fatalf("thing should mint one brand, got %v", got)
	}
	if !bt.Lookup("thing").Minted {
		t.Errorf("thing should be minted (head of the chain)")
	}
	pairedArgs := bt.Lookup("paired").PremiseArgs
	if len(pairedArgs) != 2 || len(pairedArgs[0]) != 1 || len(pairedArgs[1]) != 1 {
		t.Fatalf("paired premises should both be branded: %v", pairedArgs)
	}
	if pairedArgs[0][0] != pairedArgs[1][0] {
		t.Errorf("paired premises should unify to one brand, got %v", pairedArgs)
	}
	if got := len(bt.Params("paired")); got != 1 {
		t.Errorf("paired should have 1 surviving brand, got %d", got)
	}
}

// TestBrandUnificationViaVerifiedPremise covers rule 2's other half:
// two premises of unrelated types that a verified premise relates.
func TestBrandUnificationViaVerifiedPremise(t *testing.T) {
	bt := brandTableFor(t, `
(datatype left
  X : string;
  ==============
  [X] : left;)

(datatype right
  Y : string;
  ==============
  [Y] : right;)

(datatype joined
  L : left;
  R : right;
  (= (head L) (head R)) : verified;
  ==============
  [L R] : joined;)
`)
	args := bt.Lookup("joined").PremiseArgs
	if len(args) != 2 || len(args[0]) != 1 || len(args[1]) != 1 {
		t.Fatalf("both premises should be branded: %v", args)
	}
	if args[0][0] != args[1][0] {
		t.Errorf("verified premise relating L and R should unify their brands: %v", args)
	}
	if got := len(bt.Params("joined")); got != 1 {
		t.Errorf("joined should have 1 brand, got %d", got)
	}
}

// TestBrandsStayDistinctWhenUnrelated is the negative case: two
// premises of the same branded type that no premise relates keep two
// brands, so the conclusion is parameterized by both. Without this,
// "the check for tx1" and "the check for tx2" would be interchangeable.
func TestBrandsStayDistinctWhenUnrelated(t *testing.T) {
	bt := brandTableFor(t, `
(datatype thing
  X : string;
  ==============
  [X] : thing;)

(datatype two-things
  A : thing;
  B : thing;
  ==============
  [A B] : two-things;)
`)
	params := bt.Params("two-things")
	if len(params) != 2 {
		t.Fatalf("two unrelated thing premises should survive as 2 brands, got %v", params)
	}
	args := bt.Lookup("two-things").PremiseArgs
	if args[0][0] == args[1][0] {
		t.Errorf("unrelated premises must not share a brand: %v", args)
	}
}

// TestPrimitiveWrappersAreNotBranded pins the decision that value
// wrappers stay unparameterized.
func TestPrimitiveWrappersAreNotBranded(t *testing.T) {
	bt := brandTableFor(t, `
(datatype plain
  X : string;
  ==============
  X : plain;)

(datatype constrained-thing
  X : number;
  (>= X 0) : verified;
  ==============
  X : constrained-thing;)
`)
	for _, n := range []string{"plain", "constrained-thing"} {
		if bt.IsBranded(n) {
			t.Errorf("%s should be unbranded, got %v", n, bt.Params(n))
		}
	}
}

// TestUnconsumedCompositeIsNotBranded pins account-state's shape: a
// composite nobody reasons about needs no brand, so the host program is
// not infected with type parameters for nothing.
func TestUnconsumedCompositeIsNotBranded(t *testing.T) {
	bt := brandTableFor(t, `
(datatype leaf
  X : string;
  ==============
  X : leaf;)

(datatype lonely
  A : leaf;
  B : leaf;
  ==============
  [A B] : lonely;)
`)
	if bt.IsBranded("lonely") {
		t.Errorf("unconsumed composite should be unbranded, got %v", bt.Params("lonely"))
	}
}

// TestSumTypeBrandsPropagateToVariants checks the generic-interface
// lowering: the interface and every variant carry the same arity, so
// `HumanPrincipal[B]` satisfies `AuthenticatedPrincipal[B]`.
func TestSumTypeBrandsPropagateToVariants(t *testing.T) {
	bt := brandTableFor(t, `
(datatype inner
  X : string;
  ==============
  [X] : inner;)

(datatype variant-a
  I : inner;
  ==============
  I : either;)

(datatype variant-b
  X : string;
  ==============
  X : either;)

(datatype consumer
  E : either;
  Ok : boolean;
  (= Ok true) : verified;
  ==============
  [E Ok] : consumer;)
`)
	if n := len(bt.Params("either")); n != 1 {
		t.Fatalf("sum type should carry 1 brand, got %d", n)
	}
	for _, v := range []string{"variant-a", "variant-b"} {
		if n := len(bt.Params(v)); n != 1 {
			t.Errorf("variant %s should carry the interface's brand arity, got %d", v, n)
		}
	}
	if n := len(bt.Params("consumer")); n != 1 {
		t.Errorf("consumer should inherit the sum type's brand, got %d", n)
	}
}

// TestInferBrandsIsDeterministic guards against map-iteration order
// leaking into the table (the emitters are byte-compared by the audit
// gate, so this matters).
func TestInferBrandsIsDeterministic(t *testing.T) {
	spec, err := os.ReadFile(filepath.Join("..", "..", "examples", "multi-tenant-api", "specs", "core.shen"))
	if err != nil {
		t.Skipf("spec not readable: %v", err)
	}
	first := brandTableFor(t, string(spec)).String()
	for i := 0; i < 20; i++ {
		if got := brandTableFor(t, string(spec)).String(); got != first {
			t.Fatalf("brand table is not deterministic on run %d", i)
		}
	}
}

File: cmd/shengen-ts/brand_inference.test.ts
// Tests for the TypeScript GDP-brand inference and emission (W1).
//
// Mirrors cmd/shengen/brand_inference_test.go and
// cmd/shengen/brand_emit_test.go: the same golden tables (rendered with
// TypeScript's `<B>` instead of Go's `[B]`), the same rule-level unit
// tests, and emission assertions for the paired constructor signature,
// the witness placement, and the unbranded-output guarantee.

import { test } from "node:test";
import assert from "node:assert/strict";
import { readFileSync } from "node:fs";
import { join } from "node:path";

import { SymbolTable, generateTs, parseSpecString } from "./shengen.ts";
import { BrandTable, inferBrands } from "./brand_inference.ts";

function brandTableFor(spec: string): BrandTable {
  const parsed = parseSpecString(spec);
  const st = new SymbolTable();
  st.build(parsed.datatypes);
  st.registerDefines(parsed.defines);
  return inferBrands(parsed.datatypes, st);
}

function specFile(...parts: string[]): string {
  return readFileSync(join(import.meta.dirname, "..", "..", ...parts), "utf8");
}

function assertGolden(got: string, want: string): void {
  const normalize = (s: string) =>
    s
      .trim()
      .split("\n")
      .map((l) => l.trim().split(/\s+/).join(" "))
      .filter((l) => l !== "")
      .join("\n");
  assert.equal(normalize(got), normalize(want));
}

function generateBoth(spec: string): { plain: string; branded: string } {
  const parsed = parseSpecString(spec);
  const build = () => {
    const st = new SymbolTable();
    st.build(parsed.datatypes);
    st.registerDefines(parsed.defines);
    return st;
  };
  const st1 = build();
  const plain = generateTs(parsed.datatypes, st1, "t.shen");
  const st2 = build();
  const branded = generateTs(parsed.datatypes, st2, "t.shen", {
    brands: inferBrands(parsed.datatypes, st2),
  });
  return { plain, branded };
}

const PAYMENT_SPEC = `(datatype account-id
  X : string;
  ==============
  X : account-id;)

(datatype amount
  X : number;
  (>= X 0) : verified;
  ====================
  X : amount;)

(datatype transaction
  Amount : amount;
  From : account-id;
  To : account-id;
  ===================================
  [Amount From To] : transaction;)

(datatype balance-invariant
  Bal : number;
  Tx : transaction;
  (>= Bal (head Tx)) : verified;
  =======================================
  [Bal Tx] : balance-checked;)

(datatype safe-transfer
  Tx : transaction;
  Check : balance-checked;
  =============================
  [Tx Check] : safe-transfer;)`;

// ============================================================================
// Golden brand tables — same rows as the Go implementation
// ============================================================================

test("brand table: payment spec golden", () => {
  const bt = brandTableFor(
    specFile("examples", "payment", "specs", "core.shen")
  );
  assertGolden(
    bt.toString(),
    `
account-id               unbranded  AccountId
account-state            unbranded  AccountState  premises(- -)
amount                   unbranded  Amount
balance-checked          inherited  BalanceChecked<B>  premises(- B)
safe-transfer            inherited  SafeTransfer<B>  premises(B B)
transaction              minted     Transaction<B>  premises(- - -)
`
  );
});

test("brand table: multi-tenant spec golden", () => {
  const bt = brandTableFor(
    specFile("examples", "multi-tenant-api", "specs", "core.shen")
  );
  assertGolden(
    bt.toString(),
    `
authenticated-principal  sum        AuthenticatedPrincipal<B>
authenticated-user       inherited  AuthenticatedUser<B>  premises(B -)
human-principal          inherited  HumanPrincipal<B>  premises(B)
jwt-audience             unbranded  JwtAudience
jwt-issuer               unbranded  JwtIssuer
parsed-claims            minted     ParsedClaims<B>  premises(- - - -)
resource-access          inherited  ResourceAccess<B>  premises(B - -)
resource-id              unbranded  ResourceId
service-credential       minted     ServiceCredential<B>  premises(- -)
service-id               unbranded  ServiceId
service-principal        inherited  ServicePrincipal<B>  premises(B)
tenant-access            inherited  TenantAccess<B>  premises(B - -)
tenant-id                unbranded  TenantId
user-id                  unbranded  UserId
verified-jwt             inherited  VerifiedJwt<B>  premises(B -)
`
  );
});

test("brand table: a (list X) premise carries X's brand", () => {
  const bt = brandTableFor(
    specFile("examples", "multilang-paired", "specs", "core.shen")
  );
  assertGolden(
    bt.toString(),
    `
cart                     inherited  Cart<B>  premises(- - B)
cart-item                minted     CartItem<B>  premises(- -)
customer-id              unbranded  CustomerId
discount-eligible        inherited  DiscountEligible<B>  premises(B - - -)
sku                      unbranded  Sku
`
  );
});

// ============================================================================
// Rule-level unit tests
// ============================================================================

test("brands unify across a nested field path", () => {
  const bt = brandTableFor(PAYMENT_SPEC);
  const args = bt.lookup("safe-transfer")!.premiseArgs;
  assert.equal(args.length, 2);
  assert.equal(args[0][0], args[1][0], "Tx and Check must share a brand");
  assert.equal(bt.params("safe-transfer").length, 1);
  assert.equal(bt.lookup("transaction")!.minted, true);
  assert.equal(bt.lookup("balance-checked")!.minted, false);
});

test("brands unify when a verified premise mentions both premises", () => {
  const bt = brandTableFor(`(datatype left
  X : string;
  ==============
  [X] : left;)

(datatype right
  Y : string;
  ==============
  [Y] : right;)

(datatype joined
  L : left;
  R : right;
  (= (head L) (head R)) : verified;
  ==============
  [L R] : joined;)`);
  const args = bt.lookup("joined")!.premiseArgs;
  assert.equal(args[0][0], args[1][0]);
  assert.equal(bt.params("joined").length, 1);
});

test("unrelated premises of the same type keep distinct brands", () => {
  const bt = brandTableFor(`(datatype thing
  X : string;
  ==============
  [X] : thing;)

(datatype two-things
  A : thing;
  B : thing;
  ==============
  [A B] : two-things;)`);
  assert.equal(bt.params("two-things").length, 2);
  const args = bt.lookup("two-things")!.premiseArgs;
  assert.notEqual(args[0][0], args[1][0]);
});

test("value wrappers and unconsumed composites stay unbranded", () => {
  const bt = brandTableFor(PAYMENT_SPEC);
  for (const n of ["account-id", "amount"]) {
    assert.equal(bt.isBranded(n), false, `${n} should be unbranded`);
  }
  const lonely = brandTableFor(`(datatype leaf
  X : string;
  ==============
  X : leaf;)

(datatype lonely
  A : leaf;
  B : leaf;
  ==============
  [A B] : lonely;)`);
  assert.equal(lonely.isBranded("lonely"), false);
});

test("inferBrands is deterministic", () => {
  const spec = specFile("examples", "multi-tenant-api", "specs", "core.shen");
  const first = brandTableFor(spec).toString();
  for (let i = 0; i < 10; i++) {
    assert.equal(brandTableFor(spec).toString(), first);
  }
});

// ============================================================================
// Emission
// ============================================================================

test("brands off: output is byte-identical to the pre-brand emitter", () => {
  for (const parts of [
    ["examples", "payment", "specs", "core.shen"],
    ["examples", "multi-tenant-api", "specs", "core.shen"],
  ]) {
    const spec = specFile(...parts);
    const parsed = parseSpecString(spec);
    const st = new SymbolTable();
    st.build(parsed.datatypes);
    st.registerDefines(parsed.defines);
    const viaOldPath = generateTs(parsed.datatypes, st, "t.shen");

    const st2 = new SymbolTable();
    st2.build(parsed.datatypes);
    st2.registerDefines(parsed.defines);
    const viaNullBrands = generateTs(parsed.datatypes, st2, "t.shen", {
      brands: null,
    });

    assert.equal(viaOldPath, viaNullBrands, parts.join("/"));
    assert.ok(!viaOldPath.includes("_witness"), "witness leaked into unbranded output");
    assert.ok(!viaOldPath.includes("extends Brand"), "brand leaked into unbranded output");
  }
});

test("branded emission: paired constructor signature", () => {
  const { branded } = generateBoth(PAYMENT_SPEC);
  for (const want of [
    "export type Brand = unknown;",
    "type Witness = typeof MINTED;",
    "export class Transaction<B extends Brand> {",
    "  static createOrThrow<B extends Brand>(amount: Amount, from: AccountId, to: AccountId): Transaction<B> {",
    "  private readonly _tx: Transaction<B>;",
    "  static createOrThrow<B extends Brand>(bal: number, tx: Transaction<B>): BalanceChecked<B> {",
    "  static createOrThrow<B extends Brand>(tx: Transaction<B>, check: BalanceChecked<B>): SafeTransfer<B> {",
    "export function mustSafeTransfer<B extends Brand>(tx: Transaction<B>, check: BalanceChecked<B>): SafeTransfer<B> {",
  ]) {
    assert.ok(branded.includes(want), `branded output missing:\n\t${want}\n`);
  }
});

test("branded emission: witness on every class, checked in accessors and consumers", () => {
  const { branded } = generateBoth(PAYMENT_SPEC);
  for (const cls of ["AccountId", "Amount", "Transaction", "BalanceChecked", "SafeTransfer"]) {
    const idx = branded.indexOf(`export class ${cls}`);
    assert.ok(idx >= 0, `no class ${cls}`);
    const body = branded.slice(idx, branded.indexOf("\n}\n", idx));
    assert.ok(
      body.includes("private readonly _witness: Witness = mint();"),
      `${cls} has no witness`
    );
  }
  for (const want of [
    'val(): number { mustBeMinted(this._witness, "Amount"); return this._v; }',
    'amount(): Amount { mustBeMinted(this._witness, "Transaction"); return this._amount; }',
    'mustBeMinted((tx as unknown as { _witness?: Witness })._witness, "Transaction");',
  ]) {
    assert.ok(branded.includes(want), `branded output missing:\n\t${want}\n`);
  }
});

test("branded emission: phantom marker makes the brand invariant", () => {
  const { branded } = generateBoth(PAYMENT_SPEC);
  // `(b: B) => B` is what stops Transaction<A> being assignable to
  // Transaction<C>; an unused type parameter would be bivariant and the
  // pairing guarantee would evaporate silently.
  assert.ok(
    branded.includes("declare private readonly _brandB: (b: B) => B;"),
    "missing invariant phantom brand marker"
  );
});

test("branded emission: sum type is a generic union over the same brand", () => {
  const spec = specFile("examples", "multi-tenant-api", "specs", "core.shen");
  const { branded } = generateBoth(spec);
  for (const want of [
    "export type AuthenticatedPrincipal<B> = HumanPrincipal<B> | ServicePrincipal<B>;",
    "  private readonly _principal: AuthenticatedPrincipal<B>;",
    "  static createOrThrow<B extends Brand>(principal: AuthenticatedPrincipal<B>, tenant: TenantId, isMember: boolean): TenantAccess<B> {",
  ]) {
    assert.ok(branded.includes(want), `branded output missing:\n\t${want}\n`);
  }
});

test("branded emission: a (list X) premise is rendered at X's brand", () => {
  const spec = specFile("examples", "multilang-paired", "specs", "core.shen");
  const { branded } = generateBoth(spec);
  assert.ok(
    branded.includes("private readonly _items: CartItem<B>[];"),
    `list field should carry the element brand:\n${branded.slice(0, 200)}`
  );
  assert.ok(
    branded.includes("export function hasBulkLine<B extends Brand>(arg0: CartItem<B>[]): boolean {"),
    "a helper over branded types should be generic over their brands"
  );
});

test("branded emission: failing factory throws instead of returning a value", () => {
  const { branded } = generateBoth(PAYMENT_SPEC);
  assert.ok(
    branded.includes("if (!(bal >= tx.amount().val())) throw new Error(`bal must be >= tx.amount()`);"),
    "guarded factory should throw on a failed premise"
  );
  // tryCreate is the only non-throwing path and it returns Error, never
  // an unminted instance.
  assert.ok(
    branded.includes("static tryCreate<B extends Brand>(bal: number, tx: Transaction<B>): BalanceChecked<B> | Error {"),
    "tryCreate should return Error, not a forged value"
  );
});
